File: debot/src/main/java/com/tomoima/debot/DebotDialog.java
package com.tomoima.debot;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.VisibleForTesting;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;

import com.tomoima.debot.adapter.DebotMenuListAdapter;
import com.tomoima.debot.strategy.DebotStrategy;

import java.util.ArrayList;


public class DebotDialog extends DialogFragment {
    private static final String TAG = "com.tomoima.debot.Debot";
    private ArrayList<DebotStrategy> debotStrategyList = new DebotStrategies().getStrategies();

    public static DebotDialog getInstance() {
        return new DebotDialog();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setCancelable(true);
        onCancel(new DialogInterface() {
            @Override
            public void cancel() {
                dismissDebugMenu();
            }

            @Override
            public void dismiss() {
                dismissDebugMenu();
            }
        });
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_debot, container);
        ListView list = (ListView) view.findViewById(R.id.debug_menu_list);
        DebotMenuListAdapter adapter = new DebotMenuListAdapter(getActivity(), debotStrategyList);
        list.setAdapter(adapter);
        list.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                debotStrategyList.get(position).startAction(getActivity());
                dismissDebugMenu();
            }
        });

        return view;
    }


    public void showDebugMenu(FragmentActivity activity) {
        FragmentManager fragmentManager = activity.getSupportFragmentManager();
        if (fragmentManager != null) {
            if (fragmentManager.findFragmentByTag(TAG) == null) {
                show(fragmentManager, TAG);
                fragmentManager.executePendingTransactions();
            }
        }
    }

    public void dismissDebugMenu() {
        if (getActivity() == null) return;
        dismissDebugMenu(getActivity().getSupportFragmentManager());
    }

    @VisibleForTesting
    void dismissDebugMenu(FragmentManager fragmentManager) {

        DebotDialog debotDialog = (DebotDialog) fragmentManager.findFragmentByTag(TAG);
        if (debotDialog != null) {
            debotDialog.onDismiss(getDialog());
        }
    }
}

File: debot/src/main/java/com/tomoima/debot/DebotDialog.kt
package com.tomoima.debot

import android.content.DialogInterface
import android.os.Bundle
import androidx.annotation.VisibleForTesting
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ListView
import com.tomoima.debot.adapter.DebotMenuListAdapter

// This class allows others to inherit
open class DebotDialog : androidx.fragment.app.DialogFragment() {
    private val TAG = "com.tomoima.debot.Debot"
    private val debotStrategyList = DebotStrategies.get()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        isCancelable = true
        onCancel(object : DialogInterface {
            override fun cancel() {
                dismissDebugMenu()
            }

            override fun dismiss() {
                dismissDebugMenu()
            }
        })
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_debot, container)
        val list = view.findViewById(R.id.debug_menu_list) as ListView
        activity?.let {
            val adapter = DebotMenuListAdapter(it, debotStrategyList)
            list.adapter = adapter
            list.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
                debotStrategyList[position].startAction(it)
                dismissDebugMenu()
            }
        }


        return view
    }


    fun showDebugMenu(activity: FragmentActivity) {
        val fragmentManager = activity.supportFragmentManager
        if (fragmentManager.findFragmentByTag(TAG) == null) {
                show(fragmentManager, TAG)
                fragmentManager.executePendingTransactions()
        }
    }

    fun dismissDebugMenu() {
        activity?.let { dismissDebugMenu(it.supportFragmentManager) }
    }

    @VisibleForTesting
    fun dismissDebugMenu(fragmentManager: FragmentManager) {

        val debotDialog = fragmentManager.findFragmentByTag(TAG) as? DebotDialog ?: return
        dialog?.let(debotDialog::onDismiss)
    }

    companion object {
        fun get(): DebotDialog = DebotDialog()
    }
}
